### DISCUSSION TITLE
algorithms/ads/2

### DISCUSSION THREAD

**Discussion by giscus** [NoughtQ/notebook#15] [Announcements]
# algorithms/ads/2

NoughtQ的笔记本

https://note.noughtq.top/algorithms/ads/2

**Comment by XuebaStudy** [comment]
请问要是Red-Black Tree的Insertion的case 3中，2的"uncle"（即14）为红时（为保证性质，可使14有两个黑internal node孩子），是不是要特殊讨论（否则”调整图“最后G与D均为红）？

**Reply by NoughtQ** [reply]
你的假设应该属于case 1吧（因为此时满足自身(2)、parent(7)和uncle(14)都是红色的情况），那么只需要变个色就行了（7和14染黑，11染红后再染黑（因为发现它是根节点））

**Reply by XuebaStudy** [reply]
好的，感谢回复，我的问题😭

**Comment by Valkqs** [comment]
1.2.2 deletion处的“分析”中的“case1”的“旋转后“，图片下面的文字是不是应该是”将兄弟节点 P 染红，父节点 S 染黑”🤔

**Reply by XuebaStudy** [reply]
是“将兄弟节点S染黑，父节点P染红" 嘛

**Reply by NoughtQ** [reply]
@XuebaStudy 的表述是对的，昨天我刚复习这块，结果竟然没发现这个笔误，稍后我会修改这处错误，感谢各位指出orz（~~我到底在复习个什么~~）

**Comment by kenqia** [comment]
红黑树按理来说不应该还要存储指向父节点的指针，所以空间上应该占用更大吧

**Reply by kenqia** [reply]
那个红黑树的case状态图也有一点小问题，case2有可能能进入case1再进入case2，但是不会无限递归，这两步就终止了

**Reply by kenqia** [reply]
哦对了我上面说的是红黑树删除，红黑树删除case3的说明有一点笔误应该是L是左孩子而不是X
